services/sharedchannel/msg.go

```diff
@@ -56,3 +103,92 @@ func (scs *Service) postsToMsg(posts []*model.Post, cache msgCache) (model.Remot
 	msg := model.NewRemoteClusterMsg(TopicSync, json)
 	return msg, nil
 }
+
+// usersForPost provides a list of Users associated with the post that need to be synchronized.
+func (scs *Service) usersForPost(post *model.Post, reactions []*model.Reaction, rc *model.RemoteCluster) []*model.User {
+	userIds := make(map[string]struct{}) // avoid duplicates
+
+	if post != nil {
+		userIds[post.UserId] = struct{}{}
+	}
+
+	for _, r := range reactions {
+		userIds[r.UserId] = struct{}{}
+	}
+
+	// TODO: extract @mentions to local users and sync those as well?
+
+	users := make([]*model.User, 0)
+
+	for id := range userIds {
+		user, err := scs.server.GetStore().User().Get(id)
+		if err == nil {
+			if sync, err := scs.shouldUserSync(user, rc); err != nil {
+				continue
+			} else if sync {
+				user = sanitizeUserForSync(user)
+				users = append(users, user)
+			}
+		}
+	}
+	return users
+}
+
+func sanitizeUserForSync(user *model.User) *model.User {
+	user.Password = model.NewId()
+	user.AuthData = nil
+	user.AuthService = ""
+	user.Roles = "system_user"
+	user.AllowMarketing = false
+	user.Props = model.StringMap{}
+	user.NotifyProps = model.StringMap{}
+	user.LastPasswordUpdate = 0
+	user.LastPictureUpdate = 0
+	user.FailedAttempts = 0
+	user.MfaActive = false
+	user.MfaSecret = ""
+
+	return user
+}
+
+// shouldUserSync determines if a user needs to be synchronized.
+// User should be synchronized if it has no entry in the SharedChannelUsers table,
+// or there is an entry but the LastSyncAt is less than user.UpdateAt
+func (scs *Service) shouldUserSync(user *model.User, rc *model.RemoteCluster) (bool, error) {
+	// don't sync users with the remote they originated from.
+	if user.RemoteId != nil && *user.RemoteId == rc.RemoteId {
+		return false, nil
+	}
+
+	scu, err := scs.server.GetStore().SharedChannel().GetUser(user.Id, rc.RemoteId)
+	if err != nil {
+		if _, ok := err.(errNotFound); !ok {
+			return false, nil
+		}
+
+		// user not in the SharedChannelUsers table, so we must add them.
+		scu = &model.SharedChannelUser{
+			UserId:          user.Id,
+			RemoteClusterId: rc.RemoteId,
+		}
+		if _, err = scs.server.GetStore().SharedChannel().SaveUser(scu); err != nil {
+			scs.server.GetLogger().Log(mlog.LvlSharedChannelServiceError, "Error adding user to shared channel users",
+				mlog.String("remote_id", rc.RemoteId),
+				mlog.String("user_id", user.Id))
+		}
```

Comment:
Should we `return false, err` here instead of marking the user for sync without a `SharedChannelUser` entry?

Reply:
If we don't sync the user then the post create at the remote side will fail (assuming the user doesn't already exist there).  

Allowing the user to sync here just means we'll try to create the `SharedChannelUser` record again next time.  That table is really about not sync'ing users redundantly for every post they make, however it's better to sync them too often than not enough.